cores/esp8266/uart.c

```diff
@@ -739,3 +740,39 @@ uart_get_debug()
 {
     return s_uart_debug_nr;
 }
+
+/*
+To start detection of baud rate with the UART the UART_AUTOBAUD_EN bit needs to be cleared and set. The ROM function uart_baudrate_detect() does this only once, so on a next call the UartDev.rcv_state is not equal to BAUD_RATE_DET. Instead of poking around in the UartDev struct with unknown effect, the UART_AUTOBAUD_EN bit is directly triggered by the function uart_detect_baudrate().
+*/
+int
+uart_detect_baudrate(int uart_nr)
+{
+    static bool doTrigger = true;
+
+    if(doTrigger)
+    {
+        USA(uart_nr) &= ~(UART_GLITCH_FILT << UART_GLITCH_FILT_S | UART_AUTOBAUD_EN);
+        USA(uart_nr) = 0x08 << UART_GLITCH_FILT_S | UART_AUTOBAUD_EN;
+        doTrigger = false;
+    }
+
+    int32_t divisor = uart_baudrate_detect(uart_nr, 1);
+    if(!divisor) return 0;
+
+    doTrigger = true;    // Initialize for a next round
+    int32_t baudrate = UART_CLK_FREQ / divisor;
+
+    static const int default_rates[] = {300, 600, 1200, 2400, 4800, 9600, 19200, 38400, 57600, 74880, 115200, 230400, 256000, 460800, 921600, 1843200, 3686400};
```

Comment:
Are all of these rates valid for both 80MHz and 160MHz CPU speed?
Or rather, has the detection process been tested with both CPU speeds and different baudrates?

Reply:
These baud rates are common rates. Yes it has been tested with both 80MHz and 160 MHz, and with two devices, one at 9600 baud and one at 115200 baud. I did not test changing the crystal frequency, because I do not know the purpose of changing the crystal frequency.